session/cookie.js

```diff
@@ -7,146 +7,183 @@
 
 'use strict';
 
+const cookie = require('cookie')
+
 /**
- * Module dependencies.
+ * @import {CookieOptions} from "./cookie-options"
  */
 
-var cookie = require('cookie')
-var deprecate = require('depd')('express-session')
-
 /**
- * Initialize a new `Cookie` with the given `options`.
- *
- * @param {IncomingMessage} req
- * @param {Object} options
- * @api private
+ * Cookie TODO: add description
+ * @class
+ * @implements CookieOptions
  */
 
-var Cookie = module.exports = function Cookie(options) {
-  this.path = '/';
-  this.maxAge = null;
-  this.httpOnly = true;
+class Cookie {
+  /** @type {Date | undefined} @private */
+  _expires;
+  /** @type {number | undefined} */
+  originalMaxAge;
+  /** @type {boolean | undefined} */
+  partitioned;
+  /** @type { "low" | "medium" | "high" | undefined} */
+  priority;
+  /** @type {boolean | undefined} */
+  signed; // FIXME: how this is used??
+  /** @type {boolean} */
+  httpOnly;
+  /** @type {string} */
+  path;
+  /** @type {string | undefined} */
+  domain;
+  /** @type {boolean | "auto" | undefined} */
+  secure;
+  /** @type {((val: string) => string) | undefined} */
+  encode;
+  /** @type {boolean | "lax" | "strict" | "none" | undefined} */
+  sameSite;
 
-  if (options) {
-    if (typeof options !== 'object') {
-      throw new TypeError('argument options must be a object')
-    }
-
-    for (var key in options) {
-      if (key !== 'data') {
-        this[key] = options[key]
+  /**
+   * Initialize a new `Cookie` with the given `options`.
+   * @param {CookieOptions} options
+   * @private
+   */
+  constructor(options) {
+    if (options) {
+      if (typeof options !== 'object') {
+        throw new TypeError('argument options must be a object')
       }
+      console.log(`CookieOptions: ${JSON.stringify(options)}`)
+      this.maxAge = options.maxAge
+      this.originalMaxAge ??= options.maxAge // FIXME: rethink this
+
+      this.partitioned = options.partitioned
+      this.priority = options.priority
+      this.secure = options.secure
+      this.httpOnly = options.httpOnly ?? true
+      this.domain = options.domain
+      this.path = options.path || '/'
+      this.sameSite = options.sameSite
```

Comment:
Using a for loop here is not good from the type checking point of view (the for loop doesn't understand what type keys are).

The `Object.assign(this, options)` does work here, but then we need to skip `options.data` property some how. For some reason, people have been passing it here and it fails as `data` is readonly for `Cookie`.

As a good result, this fixes the problem with `data` field, which was special case. In addition, the code is quite clean.

As a down side, now these properties are hardcoded in quite many places and changes to those require care.

Reply:
For the second iteration, I used the old code, just added type overwrites, so that the for loop works. I think the outdated code is kind of better, but I prefer the less rewriting style at this point.